### PR TITLE
[Jenkins-46894] Adding triggeredBy when condition

### PR DESCRIPTION
**Pull request by joseblas**
* JENKINS issue(s):
    *https://issues.jenkins-ci.org/browse/JENKINS-46894
* Description:
    * Provide means of detecting cause of build from declarative pipeline
* Documentation changes:
    * https://github.com/jenkins-infra/jenkins.io/pull/1925
* Users/aliases to notify:
    * N/A

### CODE REVIEW

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditionalScript.groovy [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+class TriggeredByConditionalScript extends DeclarativeStageConditionalScript<TriggeredByConditional> {
+
+    private String cause
+    private String triggerClass
+
+    TriggeredByConditionalScript(CpsScript s, TriggeredByConditional c) {
+        super(s, c)
+        this.cause = c.cause
+        RunWrapper currentBuild = (RunWrapper)script.getProperty("currentBuild")
+        this.triggerClass = currentBuild.getBuildCauses().get(0)._class
```
[the remark this answers is not in the document]
Comment (joseblas) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditionalScript.groovy]: I did push a new version, could you take a look?

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 2]
```diff
@@ -292,6 +293,23 @@ class Utils {
         return cause != null
     }
 
+    /**
+     * Check if this run was caused by a cause.
+     */
+    @Whitelisted
```
Comment (abayer) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: Turns out we don't need these `@Whitelisted` annotations - they're from the long ago before plugin-provided Pipeline code became exempt from whitelist checks, so might as well remove this.

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditional/config.jelly [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  ~ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  ~ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  ~ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  ~ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  ~ THE SOFTWARE.
+  -->
+
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <f:entry field="compare" title="Build triggered cause">
```
Comment (dwnusbaum) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditional/config.jelly]: Does this work? It seems like the field name would be something else, but I'm not sure.

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditional/help.html [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  ~ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  ~ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  ~ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  ~ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  ~ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  ~ THE SOFTWARE.
+  -->
+
+<p>
+    Execute the stage if the <code>cause</code> variable matches at least one the reason(s) or cause(s) behind the build
```
Comment (dwnusbaum) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditional/help.html]: nit: Small typo missing `of`, and I'd change `behind` to `for` for style: 
```suggestion
    Execute the stage if the <code>cause</code> variable matches at least one of the reason(s) or cause(s) for the build
```

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditional/config.jelly (hunk 3 above)
Comment (abayer) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/TriggeredByConditional/config.jelly]: Oh d'oh, yeah, that should be `cause`, not `compare`. Lemme know when that's fixed, @joseblas!

**Pull request merged**